subt_ign/src/level_generator.cc

```diff
@@ -0,0 +1,313 @@
+using namespace subt;
+
+/// \brief Data structure containing properties of a level
+struct Level
+{
+  /// \brief Position X
+  double x;
+
+  /// \brief Position Y
+  double y;
+
+  /// \brief Position Z
+  double z;
+
+  /// \brief Size x
+  double sx;
+
+  /// \brief Size y
+  double sy;
+
+  /// \brief Size z
+  double sz;
+
+  /// \brief Buffer of a level
+  double buffer;
+
+  /// \brief A list of models referenced by this level
+  std::vector<std::string> refs;
+};
+
+/// \brief A list of models to ignore when generating levels
+const std::set<std::string> modelsToIgnore =
+{
+  "staging_area", "base_station", "artifact_origin"
+};
+
+/// \brief Set to true to print debug visuals
+bool debugLevels = false;
+
+/// \brief Print debug visuals the levels sdf string
+/// \param[in] _levels List of levels to debug
+void printDebugVisuals(const std::map<double, Level> &_levels)
+{
+  std::stringstream out;
+  static int debugCounter = 0;
+
+  for (const auto &lIt : _levels)
+  {
+    auto l = lIt.second;
+    out << "    <model name=\"level_debug_" << std::to_string(debugCounter++)
+        << "\">\n";
+    out << "      <pose>" << l.x << " " << l.y << " " << l.z
+                          << " 0 0 0</pose>\n";
+    out << "      <link name=\"level_debug_link\">\n";
+    out << "        <visual name=\"level_debug_visual\">\n";
+    out << "          <transparency>0.2</transparency>\n";
+    out << "          <geometry><box><size>\n";
+    out << "            " << l.sx << " " << l.sy << " " << l.sz << "\n";
+    out << "          </size></box></geometry>\n";
+    out << "          <material><diffuse>1 0 0</diffuse></material>\n";
+    out << "          <cast_shadows>false</cast_shadows>\n";
+    out << "        </visual>\n";
+    out << "      </link>\n";
+    out << "      <static>true</static>\n";
+    out << "    </model>\n";
+  }
+  std::cout << out.str() << std::endl;
+}
+
+/// \brief Get the levels sdf string
+/// \param[in] _level input map of levels
+std::string levelsStr(const std::map<double, Level> &_levels)
+{
+  static int levelCounter = 0;
+  std::stringstream out;
+  for (const auto &lIt : _levels)
+  {
+    auto l = lIt.second;
+    out << "      <level name=\"level" << std::to_string(levelCounter++)
+        << "\">\n";
+
+    for (const auto &r : l.refs)
+    {
+      out << "        <ref>" << r << "</ref>\n";
+    }
+    out << "        <pose>" << l.x << " " << l.y << " " << l.z
+                            << " 0 0 0</pose>\n";
+    out << "        <geometry><box><size>" << l.sx << " " << l.sy << " " << l.sz
+                            << "</size></box></geometry>\n";
+    out << "        <buffer>" << l.buffer << "</buffer>\n";
+
+    out << "      </level>\n";
+  }
+  return out.str();
+}
+
+
+/// \brief Print the level sdf elements
+/// \param[in] _vertexData vector of model info
+///\ param[in[ _size Tile size
+///\ param[in[ _buffer Level bfufer size
+void printLevels(std::vector<VertexData> &_vertexData,
```

Comment:
I think you can simplify the computation of `min`, `max`, `ls` and `l`:

```
void printLevels(std::vector<VertexData> &_vertexData,
    const math::Vector3d &_size, double _buffer)
{
  std::map<double, Level> levelX;
  std::map<double, Level> levelY;

  math::Vector3d min(math::MAX_D, math::MAX_D, math::MAX_D);
  math::Vector3d max(-math::MAX_D, -math::MAX_D, -math::MAX_D);

  // get min & max model position values
  for (const auto &vd : _vertexData)
  {
    min.Min(vd.model.Pose().Pos());
    max.Max(vd.model.Pose().Pos());
  }

  // compute level size and position
  math::Vector3d ls = max - min;
  math::Vector3d l = min + ls * 0.5;

  ls += _size * 2.0;

  double levelSizeFactor = 1.875;

  for (const auto &vd : _vertexData)
  {
    std::string name = vd.tileName;

    double x = vd.model.Pose().Pos().X();
    double y = vd.model.Pose().Pos().Y();

    // Create a new level for each unique x
    if (levelX.find(x) == levelX.end())
    {
      double rx = x;
      double ry = l.Y();
      double rz = l.Z();

      double rsx = _size.X() * levelSizeFactor;
      double rsy = ls.Y();
      double rsz = ls.Z();

      Level level;
      level.x = rx;
      level.y = ry;
      level.z = rz;
      level.sx = rsx;
      level.sy = rsy;
      level.sz = rsz;
      level.buffer = _buffer;
      levelX[x] = level;
    }
    levelX[x].refs.push_back(name);

    // Create a new level for each unique y
    if (levelY.find(y) == levelY.end())
    {
      double cx = l.X();
      double cy = y;
      double cz = l.Z();

      double csx = ls.X();
      double csy = _size.Y() * levelSizeFactor;
      double csz = ls.Z();

      Level level;
      level.x = cx;
      level.y = cy;
      level.z = cz;
      level.sx = csx;
      level.sy = csy;
      level.sz = csz;
      level.buffer = _buffer;
      levelY[y] = level;
    }
    levelY[y].refs.push_back(name);
  }
  
  // print out plugin with all the levels
  std::stringstream out;
  out << "    <!-- Auto generated: level_generator "
      << _size << " " << _buffer << " -->\n";
  out << "    <plugin name=\"ignition::gazebo\" filename=\"dummy\">\n";
  out << levelsStr(levelX);
  out << levelsStr(levelY);
  out << "    </plugin>";
  std::cout << out.str() << std::endl;

  
  // print debug visuals
  if (debugLevels)
  { 
    printDebugVisuals(levelX);
    printDebugVisuals(levelY);
  }
}
```

Reply:
updated thanks. 500847a9e4f22861f0c737b3386ae60364523fb8